### PR TITLE
[SYCL][COMPAT] First sycl_ext_oneapi_compat implementation

### PR DESCRIPTION
**Pull request by Alcpz**
This is an implementation of the experimental extension proposed in https://github.com/intel/llvm/pull/9646.

The compat extension has two primary goals:

- Improve the adoption of SYCL. The compat extension is designed to provide a familiar programming interface that resembles other popular heterogeneous programming models. By reducing the learning curve, it enables developers to leverage SYCL's power and features more easily.

- Source-to-Source Translation Support. compat is also designed to facilitate automatic source-to-source translation from other heterogeneous programming models to SYCL and offer a more standardized and consistent programming interface. This feature can significantly streamline the migration and integration of existing codebases into the SYCL ecosystem.

The PR also includes testing for the `compat` extension.

As we stated in the proposal PR, we are open to any suggestions, concerns, or improvements you may have, so please, let us know if you have any.

### CODE REVIEW

**Comment by Alcpz** [comment]
We think that we should refactor our tests from `unittests` to `e2e` tests, the reasoning being that we are targeting device code.

Our tests are built upon `gtests`, so we are wondering:
- Can we use `gtests` inside `e2e`, or should we completely avoid external dependencies there?
- Is there any possibility to have a different path, or a different build mechanism, to include the tests included in this PR?

**Comment by cperkinsintel** [comment]
Extensions should have docs. ( llvm/sycl/extensions/experimental/* ) , especially one as extensive as this.  Also, I'd like to better understand the exact use cases.  The code here seems fine, but what with wrapping atomics, kernels, and even dimensions, I'd like to get a clear understanding of what this is and how/when a user would use it.

**Comment by aelovikov-intel** [comment]
> Can we use gtests inside e2e, or should we completely avoid external dependencies there?

I think we shouldn't. Not because it's an external dependency, but because both LIT and GTest are testing frameworks and it's a bad idea to mix usage of two.

I *think* what you're looking for is `sycl/unittest-e2e` with some extra infrastructure to enable on-device testing. But I don't have any particular image in mind on how that would work.

File: sycl/include/sycl/ext/oneapi/experimental/compat.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ **************************************************************************/
+
+#pragma once
+
+#define SYCL_EXT_ONEAPI_COMPAT 1
+
+#include <sycl/ext/oneapi/experimental/compat/atomic.hpp>
+#include <sycl/ext/oneapi/experimental/compat/device.hpp>
+#include <sycl/ext/oneapi/experimental/compat/dims.hpp>
+#include <sycl/ext/oneapi/experimental/compat/dpct.hpp>
```
Comment (tomflinda) [inline sycl/include/sycl/ext/oneapi/experimental/compat.hpp:30]:
```suggestion
#include <sycl/ext/oneapi/experimental/compat/defs.hpp>
```

File: sycl/include/sycl/ext/oneapi/experimental/compat/defs.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ *  defs.hpp
+ *
+ *  Description:
+ *    helper aliases and definitions for the SYCL compatibility extension
+ *
+ **************************************************************************/
+
+// The original source was under the license below:
+//==---- dpct.hpp ---------------------------------*- C++ -*----------------==//
```
Comment (tomflinda) [inline sycl/include/sycl/ext/oneapi/experimental/compat/defs.hpp:25]:
```suggestion
//==---- defs.hpp ---------------------------------*- C++ -*----------------==//
```

**Pull request closed without merging**